p2p/protocol/identify/obsaddr.go

```diff
@@ -418,6 +434,87 @@ func (oas *ObservedAddrManager) recordObservationUnlocked(conn network.Conn, obs
 	oas.addrs[localString] = append(oas.addrs[localString], oa)
 }
 
+// For a given transport Protocol (TCP/UDP):
+//
+// 1. If we have an activated address, we are behind an Cone NAT.
+// With regards to RFC 3489, this could be either a Full Cone NAT, a Restricted Cone NAT or a
+// Port Restricted Cone NAT. However, we do NOT differentiate between them here and simply classify all such NATs as a Cone NAT.
+//
+// 2. If four different peers observe a different address for us on outbound connections, we
+// are MOST probably behind a Symmetric NAT.
+//
+// Please see the documentation on the enumerations for `network.NATDeviceType` for more details about these NAT Device types
+// and how they relate to NAT traversal via Hole Punching.
+func (oas *ObservedAddrManager) emitNATTypeUnlocked() {
+	var allObserved []*observedAddr
+	for k := range oas.addrs {
+		allObserved = append(allObserved, oas.addrs[k]...)
+	}
+
+	hasChanged, natType := oas.emitNATType(allObserved, ma.P_TCP, event.NATTransportTCP, oas.currentTCPNATDeviceType)
+	if hasChanged {
+		oas.currentTCPNATDeviceType = natType
+	}
+
+	hasChanged, natType = oas.emitNATType(allObserved, ma.P_UDP, event.NATTransportUDP, oas.currentUDPNATDeviceType)
+	if hasChanged {
+		oas.currentUDPNATDeviceType = natType
+	}
+}
+
+// returns true along with the new NAT device type if the NAT device type for the given protocol has changed.
+// returns false otherwise.
+func (oas *ObservedAddrManager) emitNATType(addrs []*observedAddr, protoCode int, transportProto event.NATTransportProtocol,
```

Comment:
This could range from a little to _very_ expensive. 
How often do we call it?
Can we skip it altogether if we are not NATted?

Reply:
@vyzo 


1. What makes this expensive ?
2. We call it every-time we have a new observed address but this function will return as soon as it sees any activated address.

The problem with skipping it altogether if we are not NATT'd is that AutoNAT dosen't take into account TCP vs UDP whereas we want to emit separate events for both.

Do you mean to say that we should emit this event for both TCP & UDP ONLY if reachability is private ?

Reply:
Imagine a big public peer with thousands of connections, this could become problematic.

And yes, this event is only relevant if we have private reachability and can/should be skipped altogether if we are public.

Reply:
done.